File: src/main/java/com/cfgglobal/test/web/api/SystemController.java
package com.cfgglobal.test.web.api;

import com.cfgglobal.test.base.ClassSearcher;
import com.cfgglobal.test.config.app.ApplicationProperties;
import com.cfgglobal.test.dao.PermissionDao;
import com.cfgglobal.test.dao.RoleDao;
import com.cfgglobal.test.domain.BaseEntity;
import com.cfgglobal.test.domain.Permission;
import com.cfgglobal.test.domain.Role;
import com.cfgglobal.test.service.GeneratorService;
import com.cfgglobal.test.service.PermissionService;
import com.cfgglobal.test.service.RoleService;
import com.google.common.collect.Lists;
import io.vavr.collection.List;
import lombok.AccessLevel;
import lombok.experimental.FieldDefaults;
import org.joor.Reflect;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import java.util.function.Consumer;

@FieldDefaults(level = AccessLevel.PRIVATE)
@RestController
@RequestMapping("/sys")
@EnableConfigurationProperties(ApplicationProperties.class)
public class SystemController {

    @Autowired
    ApplicationProperties applicationProperties;
    @Autowired
    GeneratorService userService;
    @Autowired
    PermissionService permissionService;
    @Autowired
    RoleService roleService;
    @Autowired
    PermissionDao permissionDao;
    @Autowired
    RoleDao roleDao;


    @GetMapping("/entity")
    public ResponseEntity<List<String>> entity() {
        //permissionService
        return ResponseEntity.ok(List.ofAll(ClassSearcher.of(BaseEntity.class).search()).map(e -> e.getSimpleName()));

    }

    @RequestMapping(value = "/permission", method = {RequestMethod.GET, RequestMethod.POST})
    @Transactional
    public ResponseEntity<java.util.List<Permission>> init(String entityName) {
        java.util.List<Permission> permissions = Lists.newArrayList();
        Consumer<Object> entityToPermission = e -> {
            BaseEntity entity = (BaseEntity) e;
            String name = e.getClass().getSimpleName();
            permissionDao.delete(permissionDao.findByEntity(name));
            permissions.addAll(userService.genPermission(entity).asJava());
            permissionService.save(permissions);
        };
        if (entityName != null) {
            Class clazz = Reflect.on(applicationProperties.getUserClass()).get();
            String name = clazz.getPackage().getName() + "." + entityName;
            System.out.println("entity:  " + name);
            Reflect reflect = Reflect.on(name);
            entityToPermission.accept(reflect.create().get());
        } else {
            List.ofAll(ClassSearcher.of(BaseEntity.class).search())
                    .map(e -> {
                        Reflect reflect = Reflect.on(e.getName());
                        return reflect.create().get();
                    })
                    .forEach(entityToPermission);

        }
        return ResponseEntity.ok(permissions);
    }

    @GetMapping("/assign")
    @Transactional
    public ResponseEntity<Role> assign(String roleName, String rule) {
        List<Permission> permissions = permissionService.findAll();
        Role role = roleDao.findByName(roleName).getOrElseThrow(() -> new IllegalArgumentException(roleName));
        role.getRolePermissions().clear();
        role.getRolePermissions().addAll(userService.assignPermission(permissions, rule).asJava());
        roleService.save(role);
        //TODO
        //SharedConfig.CLEAN_ROLE.end();
        return ResponseEntity.ok(role);
    }
}

File: src/main/kotlin/com/cfgglobal/test/web/api/SystemController.kt
package com.cfgglobal.test.web.api

import com.cfgglobal.test.dao.PermissionDao
import com.cfgglobal.test.dao.RoleDao
import com.cfgglobal.test.domain.BaseEntity
import com.cfgglobal.test.domain.Permission
import com.cfgglobal.test.domain.Role
import com.cfgglobal.test.service.GeneratorService
import com.cfgglobal.test.service.PermissionService
import com.cfgglobal.test.service.RoleService
import com.github.leon.classpath.ClassSearcher
import org.joor.Reflect
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/sys")
class SystemController(
        @Autowired
        val userService: GeneratorService,
        @Autowired
        val permissionService: PermissionService,
        @Autowired
        val roleService: RoleService,
        @Autowired
        val permissionDao: PermissionDao,
        @Autowired
        val roleDao: RoleDao

) {


    @GetMapping("/entity")
    fun entity(): ResponseEntity<List<String>> {
        return ResponseEntity.ok(ClassSearcher.of(BaseEntity::class.java).search<BaseEntity>().map { e -> e.simpleName })

    }

    @RequestMapping(value = ["/permission"], method = [(RequestMethod.GET), (RequestMethod.POST)])
    @Transactional
    fun init(entityName: String?): ResponseEntity<List<Permission>> {
        val permissions = mutableListOf<Permission>()
        val entityToPermission = { entity: BaseEntity ->
            val name = entity.javaClass.simpleName

            permissionDao.delete(permissionDao.findByEntity(name))
            permissions.addAll(userService.genPermission(entity))
            permissionService.save(permissions)

        }
        if (entityName != null) {
            val name = BaseEntity::class.java.`package`.name + "." + entityName
            println("entity:  " + name)
            val reflect = Reflect.on(name)
            entityToPermission(reflect.create().get<BaseEntity>())
        } else {
            ClassSearcher.of(BaseEntity::class.java).search<BaseEntity>()
                    .map { e ->
                        val reflect = Reflect.on(e.name)
                        reflect.create().get<BaseEntity>()
                    }
                    .forEach { entityToPermission }

        }
        return ResponseEntity.ok(permissions.toList())
    }

    @GetMapping("/assign")
    @Transactional
    fun assign(roleName: String, rule: String): ResponseEntity<Role> {
        val permissions = permissionService.findAll()
        val role = roleDao.findByName(roleName).getOrElseThrow { IllegalArgumentException(roleName) }
        role.rolePermissions.clear()
        role.rolePermissions.addAll(userService.assignPermission(permissions, rule))
        roleService.save(role)
        //TODO
        //SharedConfig.CLEAN_ROLE.end();
        return ResponseEntity.ok(role)
    }
}
